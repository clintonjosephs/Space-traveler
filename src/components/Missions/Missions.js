import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import fetchMissions, {
  toggleMissionSuccesful,
} from '../../redux/missions/missions';
import styles from './Missions.module.css';

let fetched = false;

const Missions = () => {
  const missions = useSelector((state) => state.missionReducer.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!fetched) {
      dispatch(fetchMissions());
      fetched = true;
    }
  }, []);

  const handleClick = (id) => {
    dispatch(toggleMissionSuccesful(id));
  };

  return (
    <table>
      <tbody>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th> </th>
        </tr>
        {missions.map(
          ({
            mission_name: missionName,
            mission_id: missionId,
            description,
            reserved,
          }) => (
            <tr key={missionId}>
              <td className={styles['mission-name']}>{missionName}</td>
              <td className={styles['mission-description']}>{description}</td>
              <td className={styles['mission-status']}>
                {' '}
                <p
                  className={
                    reserved ? styles.reserved : styles['not-reserved']
                  }
                >
                  {reserved ? 'Active Member' : 'NOT A MEMBER'}
                </p>
              </td>
              <td>
                <button
                  id={missionId}
                  onClick={(e) => handleClick(e.target.id)}
                  className={
                    reserved
                      ? styles['mission-joined']
                      : styles['mission-not-joined']
                  }
                  type="button"
                >
                  {reserved ? 'Leave Mission' : 'Join Mission'}
                </button>
              </td>
            </tr>
          ),
        )}
      </tbody>
    </table>
  );
};

export default Missions;
