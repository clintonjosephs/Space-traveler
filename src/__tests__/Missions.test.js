import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import Missions from '../components/Missions/Missions';
import store from '../redux/ConfigureStore';

const ReduxProvider = () => (
  <Provider store={store}>
    <Missions />
  </Provider>
);

describe('Missions Component', () => {
  it('renders', async () => {
    render(<ReduxProvider />);
    const missionText = screen.getByText('Mission');
    const descriptionText = screen.getByText('Description');
    const statusText = screen.getByText('Status');
    expect(missionText).toMatchSnapshot();
    expect(descriptionText).toMatchSnapshot();
    expect(statusText).toMatchSnapshot();
  });

  it('renders missions', async () => {
    render(<ReduxProvider />);
    const firstMissionName = await screen.findByText('Thaicom');
    const allMissionStatus = await screen.findAllByText('NOT A MEMBER');
    expect(firstMissionName).toMatchSnapshot();
    expect(allMissionStatus).toHaveLength(10);
  });

  it('missions change status when clicking the join mission button', async () => {
    render(<ReduxProvider />);
    const firstJoinButton = await screen.findAllByText('Join Mission');
    const firstMemberStatus = await screen.findAllByText('NOT A MEMBER');
    fireEvent.click(firstJoinButton[0]);
    expect(firstJoinButton[0].textContent).toBe('Leave Mission');
    expect(firstMemberStatus[0].textContent).toBe('Active Member');
    fireEvent.click(firstJoinButton[0]);
    expect(firstJoinButton[0].textContent).toBe('Join Mission');
    expect(firstMemberStatus[0].textContent).toBe('NOT A MEMBER');
  });
});
