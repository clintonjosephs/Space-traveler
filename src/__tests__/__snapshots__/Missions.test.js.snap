// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component renders 1`] = `
<th>
  Mission
</th>
`;

exports[`Missions Component renders 2`] = `
<th>
  Description
</th>
`;

exports[`Missions Component renders 3`] = `
<th>
  Status
</th>
`;

exports[`Missions Component renders missions 1`] = `
<td
  class="mission-name"
>
  Thaicom
</td>
`;
